planner/core/rule_eliminate_projection.go

```diff
@@ -137,6 +137,14 @@ func (pe *projectionEliminater) eliminate(p LogicalPlan, replace map[string]*exp
 		}
 	}
 	p.replaceExprColumns(replace)
+	if isProj {
+		if child, ok := p.Children()[0].(*LogicalProjection); ok && !exprsHasSideEffects(child.Exprs) {
```

Comment:
This optimization seems a little ad-hoc. I think a general way is what @eurekaka pointed out. But never mind, the current optimizer is not well suited for these kinds of optimizations.

In the cascades planner, we can push `project` operator down, merge two adjacent `project` operators when we find this pattern.